cmd/joy/execute.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/nestoca/survey/v2"
+
+	"github.com/nestoca/joy/internal/config"
+
+	"github.com/spf13/cobra"
+)
+
+func NewExecuteCmd() *cobra.Command {
+	var list bool
+	var jsonOutput bool
+	cmd := &cobra.Command{
+		Use:     "execute",
+		Aliases: []string{"exec", "x"},
+		Args:    cobra.ArbitraryArgs,
+		Short:   "Execute user joy scripts",
+		Long:    `Execute custom user joy scripts.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cfg := config.FromContext(cmd.Context())
+			searchPath := os.Getenv("PATH")
+
+			scripts, err := getScripts(cfg.CatalogDir, searchPath)
+			if err != nil {
+				return fmt.Errorf("getting scripts: %w", err)
+			}
+
+			if list {
+				if jsonOutput {
+					output, err := formatScriptsAsJson(scripts)
+					if err != nil {
+						return fmt.Errorf("formatting scripts as JSON: %w", err)
+					}
+					fmt.Println(output)
+					return nil
+				}
+
+				fmt.Println(formatScripts(scripts))
+				return nil
+			}
+
+			var script Script
+			if len(args) == 0 {
+				script, err = selectScript(scripts)
+				if err != nil {
+					return fmt.Errorf("selecting script: %w", err)
+				}
+			} else {
+				script, err = getScript(scripts, args[0])
+				if err != nil {
+					return fmt.Errorf("getting script: %w", err)
+				}
+				args = args[1:]
+			}
+
+			return execute(cfg.CatalogDir, script.Path, args)
+		},
+	}
+
+	cmd.Flags().BoolVarP(&list, "list", "l", false, "List available scripts")
+	cmd.Flags().BoolVarP(&jsonOutput, "json", "j", false, "Output as JSON")
+
+	return cmd
+}
+
+func selectScript(scripts []Script) (Script, error) {
+	var scriptNames []string
+	for _, script := range scripts {
+		scriptNames = append(scriptNames, script.Name)
+	}
+	prompt := &survey.Select{
+		Message: "Select script to execute:",
+		Options: scriptNames,
+	}
+	var index int
+	if err := survey.AskOne(prompt, &index); err != nil {
+		return Script{}, fmt.Errorf("prompting for script selection: %w", err)
+	}
+	return scripts[index], nil
+}
+
+type Script struct {
+	Name string `json:"name"`
+	Path string `json:"path"`
+}
+
+func getScripts(catalogDir, searchPath string) ([]Script, error) {
+	scriptsMap := make(map[string]Script)
+
+	catalogBinDir := filepath.Join(catalogDir, "bin")
+	files, err := os.ReadDir(catalogBinDir)
+	if err != nil {
+		return nil, fmt.Errorf("listing files under catalog bin dir %q: %w", catalogBinDir, err)
+	}
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		name := file.Name()
+		scriptsMap[name] = Script{
+			Name: name,
+			Path: filepath.Join(catalogBinDir, file.Name()),
+		}
+	}
+
+	searchDirs := filepath.SplitList(searchPath)
```

Comment:
I would recommend KISS: joy is wired to work with a catalog. Joy extensions should only live in the catalog, as the extensions are meant to extend the behaviour of that catalog.  Let's not look for scripts in the user's PATH.

Reply:
In that sense we don't even need "joy" prefixes.

Reply:
Then we don't even need a type Script, it would just be the filenames under `catalog/bin`

Reply:
I disagree, part of the goal is to allow power users to create their own scripts! 🤷

Reply:
And that loosely follows the same way that `kubectl` extensions can be defined, by defining them with the `kubectl-` prefix and placing them in your `PATH`.

Reply:
I really don't see the difference for a power user between invoking an extension like so:
```
joy exec extension
---
joy-extension
```

I am not sure the environment variable injection adds much value, and it basically encourages users to have private scripts instead of developing useful utilities in the public eye.

Less is more. More is only ever more, when it is less.

Reply:
You better check my `PATH`! ;) 

`git` also did the same thing well before k8s was a thing.

Reply:
Indeed. Though it unifies user scripts with catalog scripts, makes them discoverable with `joy exec --list`, and automatically injects the `JOY_CATALOG_DIR` env var.  And eventually we can pass more context to those scripts.